File: localstorage/src/main/java/ch/fhnw/geiger/localstorage/db/data/NodeValueImpl.java
package ch.fhnw.geiger.localstorage.db.data;

import ch.fhnw.geiger.serialization.SerializerHelper;
import ch.fhnw.geiger.totalcross.ByteArrayInputStream;
import ch.fhnw.geiger.totalcross.ByteArrayOutputStream;
import ch.fhnw.geiger.totalcross.Locale;
import ch.fhnw.geiger.totalcross.MissingResourceException;
import ch.fhnw.geiger.totalcross.System;
import java.io.IOException;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

/**
 * <p>This abstract class defines the common attributes for all NodeValueObjects.</p>
 *
 * @author Sacha
 * @version 0.1
 */
public class NodeValueImpl implements NodeValue {

  private static final long serialversionUID = 871283188L;

  private static final Locale DEFAULT_LOCALE = Locale.ENGLISH;

  /**
   * They key is used to identify the NodeValueObject inside a StorageNode,
   * therefore, the key is unique within one StorageNode.
   */
  private String key;

  private final Map<Locale, String> value = new HashMap<>();

  /**
   * <p>The type of the value.</p>>
   */
  private String type;

  /**
   * Description of this value, can be used for translation.
   */
  private final Map<Locale, String> description = new HashMap<>();

  /**
   * Defines the epoch when this value was last modified.
   */
  private long lastModified;

  /**
   * <p>Default constructor to create a new key/value pair.</p>
   *
   * <p>Type and description of the key/value pair are set to null. This
   * Constructor should only be used for private values (visibility RED).</p>
   *
   * @param key   the name of the key
   * @param value the value of the key
   */
  public NodeValueImpl(String key, String value) {
    this(key, value, null, null, 0);
  }

  /**
   * <p>A fully fledged constructor for creating key/value pairs suitable for sharing.</p>
   *
   * @param key          the name of the key/value pair
   * @param value        the value to be set
   * @param type         a searchable type field
   * @param description  the description to be shown when asking for consent of sharing
   * @param lastModified the last modified date to be set
   */
  public NodeValueImpl(String key, String value, String type, String description,
                       long lastModified) {
    if (key == null || value == null) {
      throw new NullPointerException();
    }
    this.key = key;
    setLocalizedString(this.value, value, DEFAULT_LOCALE);
    this.type = type;
    if (description != null) {
      setLocalizedString(this.description, description, DEFAULT_LOCALE);
    }
    this.lastModified = lastModified;
  }

  @Override
  public String getKey() {
    return key;
  }

  private void setKey(String key) {
    this.key = key;
    updateLastmodified();
  }

  @Override
  public String getValue() {
    return getValue(DEFAULT_LOCALE.toLanguageTag());
  }

  @Override
  public String getValue(String languageRange) {
    return getLocalizedString(this.value, languageRange);
  }

  @Override
  public Map<Locale, String> getAllValueTranslations() {
    return new HashMap<>(value);
  }

  @Override
  public void setValue(String value, Locale locale) throws MissingResourceException {
    setLocalizedString(this.value, value, locale);
    updateLastmodified();
  }

  @Override
  public void setValue(String value) {
    setValue(value, DEFAULT_LOCALE);
  }

  @Override
  public String getType() {
    return type;
  }

  @Override
  public String setType(String type) {
    String ret = this.type;
    this.type = type;
    updateLastmodified();
    return ret;
  }

  @Override
  public String getDescription() {
    return getDescription(DEFAULT_LOCALE.toLanguageTag());
  }

  @Override
  public String getDescription(String languageRange) {
    return getLocalizedString(description, languageRange);
  }

  @Override
  public Map<Locale, String> getAllDescriptionTranslations() {
    return new HashMap<>(description);
  }

  private static Locale lookupLocale(Map<Locale, String> map, String languageRange) {
    // Get Language Range
    List<Locale> matchingLocales = Locale.filter(Locale.LanguageRange.parse(languageRange),
        map.keySet());
    if (matchingLocales.size() > 0) {
      return matchingLocales.get(0);
    } else {
      return DEFAULT_LOCALE;
    }
  }

  private static String getLocalizedString(Map<Locale, String> map, String languageRange) {
    return map.get(lookupLocale(map, languageRange));
  }

  private static void setLocalizedString(Map<Locale, String> map, String value, Locale locale)
      throws MissingResourceException {
    if (getLocalizedString(map, DEFAULT_LOCALE.toLanguageTag()) == null
        && locale != DEFAULT_LOCALE) {
      throw new MissingResourceException("undefined string for locale " + DEFAULT_LOCALE, "Locale",
          locale.toLanguageTag());
    }
    map.put(locale, value);
  }

  @Override
  public String setDescription(String value, Locale locale) {
    if (description == null) {
      throw new NullPointerException("description may not be null");
    }
    String ret = getLocalizedString(this.description, locale.toLanguageTag());
    setLocalizedString(this.description, value, locale);
    updateLastmodified();
    return ret;
  }

  @Override
  public String setDescription(String description) {
    return setDescription(description, DEFAULT_LOCALE);
  }

  @Override
  public long getLastModified() {
    return lastModified;
  }

  @Override
  public void update(NodeValue node) {
    NodeValueImpl n2 = (NodeValueImpl) (node);
    this.key = n2.getKey();
    this.value.clear();
    for (Map.Entry<Locale, String> e : n2.value.entrySet()) {
      this.value.put(e.getKey(), e.getValue());
    }
    this.type = n2.getType();
    this.description.clear();
    for (Map.Entry<Locale, String> e : n2.description.entrySet()) {
      this.description.put(e.getKey(), e.getValue());
    }
    updateLastmodified();
  }

  private void updateLastmodified() {
    this.lastModified = ch.fhnw.geiger.totalcross.System.currentTimeMillis();
  }

  @Override
  public NodeValue deepClone() {
    NodeValue ret = new NodeValueImpl(getKey(), getValue());
    ret.update(this);
    return ret;
  }

  @Override
  public String toString() {
    return toString("");
  }

  /***
   * <p>prints a space prefixed representation of the NodeValue.</p>
   *
   * @param prefix a prefix (typically a series of spaces
   * @return the string representation
   */
  public String toString(String prefix) {
    StringBuilder sb = new StringBuilder();
    // build head of value
    sb.append(prefix).append(getKey());
    if (getType() != null) {
      sb.append(":").append(getType());
    }
    // build values
    sb.append("={");
    if (value.size() == 1) {
      sb.append(DEFAULT_LOCALE).append("=>\"").append(value.get(DEFAULT_LOCALE)).append("\"}");
    } else {
      sb.append(System.lineSeparator());
      int i = 0;
      for (Locale l : new TreeSet<>(value.keySet())) {
        if (i > 0) {
          sb.append(",").append(System.lineSeparator());
        }
        sb.append(prefix).append("  ").append(l.toLanguageTag()).append("=>\"")
            .append(value.get(l)).append("\"");
        i++;
      }
      sb.append(System.lineSeparator()).append(prefix).append("}");
      // build description

    }
    return sb.toString();
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof NodeValueImpl)) {
      return false;
    }
    NodeValueImpl nv = (NodeValueImpl) o;

    return toString().equals(nv.toString());
  }

  @Override
  public void toByteArrayStream(ByteArrayOutputStream out) throws IOException {
    // write object identifier
    SerializerHelper.writeLong(out, serialversionUID);

    // write key
    SerializerHelper.writeString(out, key);

    // value
    SerializerHelper.writeInt(out, value.size());
    synchronized (value) {
      for (Map.Entry<Locale, String> e : value.entrySet()) {
        SerializerHelper.writeString(out, e.getKey().toLanguageTag());
        SerializerHelper.writeString(out, e.getValue());
      }
    }

    // type
    SerializerHelper.writeString(out, type);

    // lastModified
    SerializerHelper.writeLong(out, lastModified);

    // description
    SerializerHelper.writeInt(out, description.size());
    synchronized (description) {
      for (Map.Entry<Locale, String> e : description.entrySet()) {
        SerializerHelper.writeString(out, e.getKey().toLanguageTag());
        SerializerHelper.writeString(out, e.getValue());
      }
    }

    // write object identifier as end tag
    SerializerHelper.writeLong(out, serialversionUID);

  }

  /**
   * <p>Deserializes a NodeValue from a byteStream.</p>
   *
   * @param in the stream to be read
   * @return the deserialized NodeValue
   * @throws IOException if an exception happens deserializing the stream
   */
  public static NodeValueImpl fromByteArrayStream(ByteArrayInputStream in) throws IOException {
    if (SerializerHelper.readLong(in) != serialversionUID) {
      throw new IOException("failed to parse NodeValueImpl (bad stream?)");
    }

    // read key
    NodeValueImpl nv = new NodeValueImpl(SerializerHelper.readString(in), "");

    // restore values
    int counter = SerializerHelper.readInt(in);
    nv.value.clear();
    for (int i = 0; i < counter; i++) {
      nv.value.put(Locale.forLanguageTag(SerializerHelper.readString(in)),
          SerializerHelper.readString(in));
    }

    // restore type
    nv.type = SerializerHelper.readString(in);

    // restore lastModified
    nv.lastModified = SerializerHelper.readLong(in);

    // restore description
    counter = SerializerHelper.readInt(in);
    for (int i = 0; i < counter; i++) {
      nv.description.put(Locale.forLanguageTag(SerializerHelper.readString(in)),
          SerializerHelper.readString(in));
    }

    if (SerializerHelper.readLong(in) != serialversionUID) {
      throw new IOException("failed to parse NodeValueImpl (bad stream end?)");
    }

    return nv;
  }
}

File: lib/src/db/data/NodeValueImpl.dart
import 'dart:collection';

import 'package:intl/locale.dart';

import 'NodeValue.dart';

/// <p>This abstract class defines the common attributes for all NodeValueObjects.</p>
///
/// @author Sacha
/// @version 0.1
class NodeValueImpl with NodeValue {
  static const int serialversionUID = 871283188;
  static final Locale DEFAULT_LOCALE = Locale.fromSubtags(languageCode: 'en');

  /// They key is used to identify the NodeValueObject inside a StorageNode,
  /// therefore, the key is unique within one StorageNode.
  String key;
  final Map<String, String> value = HashMap();

  /// <p>The type of the value.</p>>
  String? type;

  /// Description of this value, can be used for translation.
  final Map<String, String> description = HashMap();

  /// Defines the epoch when this value was last modified.
  late int lastModified;

  /// <p>A fully fledged constructor for creating key/value pairs suitable for sharing.</p>
  ///
  /// @param key          the name of the key/value pair
  /// @param value        the value to be set
  /// @param type         a searchable type field
  /// @param description  the description to be shown when asking for consent of sharing
  /// @param lastModified the last modified date to be set
  NodeValueImpl(this.key, String value,
      [this.type, String? description, int? lastModified]) {
    setLocalizedString(this.value, value, DEFAULT_LOCALE);
    if (description != null) {
      setLocalizedString(this.description, description, DEFAULT_LOCALE);
    }
    this.lastModified = lastModified ?? 0;
  }

  @override
  String getKey() {
    return key;
  }

  void setKey(String key) {
    this.key = key;
    updateLastModified();
  }

  @override
  String? getValue([String? languageRange]) {
    return getLocalizedString(
        value, languageRange ?? DEFAULT_LOCALE.toLanguageTag());
  }

  @override
  Map<Locale, String> getAllValueTranslations() {
    Map<Locale, String> m = HashMap();
    for (var e in value.entries) {
      m[Locale.parse(e.key)] = e.value;
    }
    return m;
  }

  @override
  void setValue(String value, [Locale? locale]) {
    setLocalizedString(this.value, value, locale ?? DEFAULT_LOCALE);
    updateLastModified();
  }

  @override
  String? getType() {
    return type;
  }

  @override
  String? setType(String type) {
    var ret = this.type;
    this.type = type;
    updateLastModified();
    return ret;
  }

  @override
  String? getDescription([String? languageRange]) {
    return getLocalizedString(
        description, languageRange ?? DEFAULT_LOCALE.toLanguageTag());
  }

  @override
  Map<Locale, String> getAllDescriptionTranslations() {
    Map<Locale, String> m = HashMap();
    for (var e in description.entries) {
      m[Locale.parse(e.key)] = e.value;
    }
    return m;
  }

  static Locale lookupLocale(Map<String, String> map, String languageRange) {
    if(map.isEmpty) {
      return DEFAULT_LOCALE;
    }
    var l = map.keys.first;
    // TODO: implement lookup with language range
    // Locale.lookupTag(Locale.LanguageRange.parse(languageRange), map.keys);
    if (l != null) {
      return Locale.parse(l);
    } else {
      return DEFAULT_LOCALE;
    }
  }

  static String? getLocalizedString(
      Map<String, String> map, String languageRange) {
    return map[lookupLocale(map, languageRange).toLanguageTag()];
  }

  static void setLocalizedString(
      Map<String, String> map, String value, Locale locale) {
    if ((getLocalizedString(map, DEFAULT_LOCALE.toLanguageTag()) == null) &&
        (!(locale.toLanguageTag() == DEFAULT_LOCALE.toLanguageTag()))) {
      throw Exception('undefined string for locale ' +
          DEFAULT_LOCALE.toString() +
          ' Locale ' +
          locale.toLanguageTag());
    }
    map[locale.toLanguageTag()] = value;
  }

  @override
  String? setDescription(String value, [Locale? locale]) {
    locale ??= DEFAULT_LOCALE;
    var ret = getLocalizedString(description, locale.toLanguageTag());
    setLocalizedString(description, value, locale);
    updateLastModified();
    return ret;
  }

  @override
  int getLastModified() {
    return lastModified;
  }

  @override
  void update(NodeValue node) {
    var n2 = node as NodeValueImpl;
    key = n2.getKey();
    value.clear();
    for (var e in n2.value.entries) {
      value[e.key] = e.value;
    }
    type = n2.getType();
    description.clear();
    for (var e in n2.description.entries) {
      description[e.key] = e.value;
    }
    updateLastModified();
  }

  void updateLastModified() {
    lastModified = DateTime.now().millisecondsSinceEpoch;
  }

  @override
  NodeValue deepClone() {
    NodeValue ret = NodeValueImpl(getKey(), getValue()!);
    ret.update(this);
    return ret;
  }

  /// <p>prints a space prefixed representation of the NodeValue.</p>
  ///
  /// @param prefix a prefix (typically a series of spaces
  /// @return the string representation
  @override
  String toString([String prefix = '']) {
    var sb = StringBuffer();
    sb.write(prefix);
    sb.write(getKey());
    if (getType() != null) {
      sb.write(':');
      sb.write(getType());
    }
    sb.write('={');
    if (value.length == 1) {
      sb.write(DEFAULT_LOCALE.toLanguageTag());
      sb.write('=>\"');
      sb.write(value[DEFAULT_LOCALE.toLanguageTag()]);
      sb.write('\"}');
    } else {
      sb.write('\n');
      var i = 0;
      var keyList = List.from(value.keys);
      keyList.sort();
      for (String l in keyList) {
        if (i > 0) {
          sb.write(',');
          sb.write('\n');
        }
        sb.write(prefix);
        sb.write('  ');
        sb.write(Locale.parse(l).toLanguageTag());
        sb.write('=>\"');
        sb.write(value[l]);
        sb.write('\"');
        i++;
      }
      sb.write('\n');
      sb.write(prefix);
      sb.write('}');
    }
    return sb.toString();
  }

  bool equals(Object o) {
    if (!(o is NodeValueImpl)) {
      return false;
    }
    return toString() == o.toString();
  }

/*void toByteArrayStream(Sink<List<int>> out) {
    SerializerHelper.writeLong(out, serialversionUID);
    SerializerHelper.writeString(out, key);
    SerializerHelper.writeInt(out, value.length);
    synchronized(value, {
    for (MapEntry<String, String> e in value.entries) {
    SerializerHelper.writeString(out, e.key);
    SerializerHelper.writeString(out, e.value);
    }
    });
    SerializerHelper.writeString(out, type);
    SerializerHelper.writeLong(out, lastModified);
    SerializerHelper.writeInt(out, description.length);
    synchronized(description, {
    for (MapEntry<String, String> e in description.entries) {
    SerializerHelper.writeString(out, e.key);
    SerializerHelper.writeString(out, e.value);
    }
    });
    SerializerHelper.writeLong(out, serialversionUID);
  }

  /// <p>Deserializes a NodeValue from a byteStream.</p>
  /// @param in the stream to be read
  /// @return the deserialized NodeValue
  /// @throws IOException if an exception happens deserializing the stream
  static NodeValueImpl fromByteArrayStream(Stream<List<int>> in_) {
    if (SerializerHelper.readLong(in_) != serialversionUID) {
      throw java_io_IOException(
          'failed to parse NodeValueImpl (bad stream?)');
    }
    var nv = NodeValueImpl(SerializerHelper.readString(in_), '');
    int counter = SerializerHelper.readInt(in_);
    nv.value.clear();
    for (var i = 0; i < counter; i++) {
      nv.value.put(Locale.forLanguageTag(SerializerHelper.readString(in_))
          .toLanguageTag(), SerializerHelper.readString(in_));
    }
    nv.type = SerializerHelper.readString(in_);
    nv.lastModified = SerializerHelper.readLong(in_);
    counter = SerializerHelper.readInt(in_);
    for (var i = 0; i < counter; i++) {
      nv.description.put(Locale.forLanguageTag(SerializerHelper.readString(in_))
          .toLanguageTag(), SerializerHelper.readString(in_));
    }
    if (SerializerHelper.readLong(in_) != serialversionUID) {
      throw java_io_IOException(
          'failed to parse NodeValueImpl (bad stream end?)');
    }
    return nv;
  }*/

}
